File: Sheets/src/neetcode/practice/Linked_List/_1_reverseLinkedList.java
package neetcode.practice.Linked_List;

class Solution {
  public ListNode reverseList(ListNode head) {
    if (head == null || head.next == null)
      return head;

    ListNode prevNode = null, currentNode = head, nextNode = head.next;
    while (nextNode != null) {
      currentNode.next = prevNode;
      prevNode = currentNode;
      currentNode = nextNode;
      nextNode = currentNode.next;
    }

    currentNode.next = prevNode;
    return currentNode;
  }
}

File: Sheets/src/neetcode/practice/Linked_List/_1_ReverseLinkedList.cc
#include "./ListNode.hpp"

class Solution {
public:
  ListNode *newHead = nullptr;

  void reverse(ListNode *head) {
    if (head->next->next == nullptr) {
      newHead = head->next;
      head->next->next = head;
      head->next = nullptr;
      return;
    }

    reverse(head->next);
    head->next->next = head;
    head->next = nullptr;
  }

  ListNode *reverseList(ListNode *head) {
    if (head == nullptr || head->next == nullptr)
      return head;

    reverse(head);
    return newHead;
  }
};
